### PR TITLE
#166266544 Add email login

### PR DESCRIPTION
**Pull request by ocranbillions**
### What does this PR do?
- POST /api/v1/auth/signin
- It creates a /auth/signin endpoint where users can sign in via email

#### Description of Task to be completed?
- Add /signin endpoint to routes
- Add validation middleware for user inputs
- Add signinController as route handler
- Get email, password from request body
- Query db for user
- Return response based on the query result

#### How should this be manually tested?
- Create a new user
- Test by signing in the user with their email and password
#### Any background context you want to provide?
- I went through @PelumiAlesh' s `signup validation branch` and applied his signup logic to my sign in endpoint since this endpoint will also need validations.
- In the userRoute.js file, I noticed @cvjude mistakenly added his socialSignIn controller to Kafiat's /signup route. I fixed that on this branch because the signup route wasn't working anymore so I couldn't signup a new user to test my sign in implementation

#### What are the relevant pivotal tracker stories?
- [166266544](https://www.pivotaltracker.com/story/show/166266544)

#### Screenshots (if appropriate)
postman

#### Questions:
- N/A

### CODE REVIEW

**Comment by Hadeneekeh** [comment]
Travis is failing

File: .env.sample [hunk 1]
```diff
@@ -1,4 +1,8 @@
 PORT = an integer representing the port through which the server will be accessed e.g. 3000
 SECRET_KEY = A string to be used when generating token, it can be any length.
 DATABASE_URL= postgres://user_name:password@host:port/db_name
-TEST_DATABASE_URL=postgres://user_name:password@host:port/test_db_name
\ No newline at end of file
+TEST_DATABASE_URL=postgres://user_name:password@host:port/test_db_name
+google_client  = string
+google_secret = string
+facebook_client = string
+facebook_secret = string
```
Comment (chuxmykel) [inline .env.sample:8]: Nice one

**Review by chuxmykel — APPROVED** [review]

File: server/controllers/userController.js [hunk 2]
```diff
@@ -82,6 +82,31 @@ class UserController {
...
+
+  static async emailSignin(req, res) {
+    try {
+      const { email, password } = req.body;
+      const user = await models.Users.findOne({ where: { email } });
+      if (!user) return util.errorstatus(res, 404, 'User not found!');
+     
+      const result = await auth.comparePassword(password, user.password);
+
+      if (!result) return util.errorstatus(res, 403, 'Incorrect password!');
```
Comment (chuxmykel) [inline server/controllers/userController.js]: I think a 401 error code is more appropriate here bro. Unauthorized as opposed to forbidden

**Review by chuxmykel — CHANGES_REQUESTED** [review]

File: server/controllers/userController.js (hunk 2 above)
Comment (ocranbillions) [reply server/controllers/userController.js]: Ok. will change that.

**Review by Hadeneekeh — DISMISSED** [review]

File: server/__test__/user.spec.js [hunk 3]
```diff
@@ -154,4 +155,64 @@ describe('User tests', () => {
...
+          expect(res.body).toMatchSnapshot();
+        });
+    });
+
+    it('Should not sign in user with wrong email', async done => {
+      server()
+        .post(`${url}/auth/signin`)
+        .send(mockUser.wrongEmail)
+        .end((err, res) => {
+          expect(res.statusCode).toEqual(404);
```
Comment (chuxmykel) [inline server/__test__/user.spec.js]: Bro here's what I think. It should be a 401 error. The user is not searching for a resource, he's trying to log in. So it should be a 401 if the email and password actually match the correct format, and if they don't match the format for instance `wrong-email@mail` as opposed to `wrong-email@mail.com`, it should be a 400 error.

**Review by chuxmykel — CHANGES_REQUESTED** [review]
I'm really sorry my reviews are coming in bits. I'm struggling with electricity.

File: server/controllers/userController.js [hunk 4]
```diff
@@ -82,6 +82,31 @@ class UserController {
       signupMethod: user.signupMethod
     });
   }
+
+  static async emailSignin(req, res) {
+    try {
+      const { email, password } = req.body;
+      const user = await models.Users.findOne({ where: { email } });
+      if (!user) return util.errorstatus(res, 404, 'User not found!');
```
Comment (chuxmykel) [inline server/controllers/userController.js]: # 401

**Review by cvjude — DISMISSED** [review]

File: server/__test__/user.spec.js (hunk 3 above)
Comment (ocranbillions) [reply server/__test__/user.spec.js]: me and these status codes. 
Thanks bro

File: server/controllers/userController.js [hunk 5]
```diff
@@ -82,6 +82,31 @@ class UserController {
       signupMethod: user.signupMethod
     });
   }
+
+  static async emailSignin(req, res) {
+    try {
+      const { email, password } = req.body;
+      const user = await models.Users.findOne({ where: { email } });
+      if (!user) return util.errorstatus(res, 401, 'Incorrect Login information');
```
Comment (encodedBicoding) [inline server/controllers/userController.js]: this was changed in my recent pr to follow the standard camelCase naming, so `util.errorstatus` is now `util.errorStatus`.

File: server/controllers/userController.js [hunk 6]
```diff
@@ -82,6 +82,31 @@ class UserController {
...
+     
+      const result = await auth.comparePassword(password, user.password);
+
+      if (!result) return util.errorstatus(res, 401, 'Incorrect Login information');
```
Comment (encodedBicoding) [inline server/controllers/userController.js]: same here

File: server/middleware/validation.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      .withMessage('Email should not be left empty: Please input email address')
+      .isEmail()
+      .trim()
+      .withMessage('Email is not valid: Please input a valid email address'),
+    check('password')
+      .not()
+      .isEmpty({ ignore_whitespace: true })
+      .withMessage('Password should not be empty: Please input password')
+      .trim()
+      .isLength({ min: 7 })
```
Comment (encodedBicoding) [inline server/middleware/validation.js:17]: standard minimum is 8, isLength has min of 7

**Review by chuxmykel — APPROVED** [review]
LGTM

**Review by Hadeneekeh — DISMISSED** [review]

**Review by encodedBicoding — CHANGES_REQUESTED** [review]

File: server/controllers/userController.js (hunk 5 above)
Comment (chuxmykel) [reply server/controllers/userController.js]: @ocranbillions Pls effect the change let's move on.

**Review by chuxmykel — CHANGES_REQUESTED** [review]

**Pull request closed without merging**